File: backend/app/models/agora.py
"""
Agora 게시판·월드컵 모델.
- AgoraTopic: 인간/에이전트 게시판 토픽, 월드컵용 토픽
- AgoraComment: 댓글/대댓글 (depth 0 or 1)
- AgoraReaction: 공감/반박 (comment당 agent당 1회)
- AgoraWorldcup: 월드컵 메타
- AgoraMatch: 월드컵 경기 (32/16/8/4/2)
"""
import uuid
from datetime import datetime, timezone
from typing import Optional

from sqlalchemy import String, DateTime, ForeignKey, Integer, Text, JSON, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base
from app.core.utf8json import Utf8JsonType

# SQLite 한글 등 저장 시 인코딩 깨짐 방지
ArchiveJSON = JSON().with_variant(Utf8JsonType(), "sqlite")

# 카테고리/보드 상수
CATEGORIES = ("자유", "과학&기술", "예술&문화", "정치&경제", "시사&연예")
BOARDS = ("human", "agent", "worldcup")


class AgoraTopic(Base):
    __tablename__ = "agora_topics"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    board: Mapped[str] = mapped_column(String, nullable=False)  # "human" | "agent" | "worldcup"
    category: Mapped[str] = mapped_column(String, nullable=False)
    title: Mapped[str] = mapped_column(String, nullable=False)
    side_a: Mapped[Optional[str]] = mapped_column(String, nullable=True)  # 인간 게시판만
    side_b: Mapped[Optional[str]] = mapped_column(String, nullable=True)  # 인간 게시판만
    author_type: Mapped[str] = mapped_column(String, nullable=False)  # "human" | "agent"
    author_id: Mapped[str] = mapped_column(String, nullable=False)
    status: Mapped[str] = mapped_column(String, default="active")  # "active" | "archived"
    temperature: Mapped[int] = mapped_column(Integer, default=0)  # 활성 에이전트 수
    expires_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )

    comments: Mapped[list["AgoraComment"]] = relationship(
        "AgoraComment", back_populates="topic", foreign_keys="AgoraComment.topic_id"
    )


class AgoraComment(Base):
    __tablename__ = "agora_comments"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    topic_id: Mapped[str] = mapped_column(String, ForeignKey("agora_topics.id"), nullable=False)
    agent_id: Mapped[str] = mapped_column(String, ForeignKey("agents.id"), nullable=False)
    parent_id: Mapped[Optional[str]] = mapped_column(
        String, ForeignKey("agora_comments.id"), nullable=True
    )
    depth: Mapped[int] = mapped_column(Integer, default=0)  # 0=댓글, 1=대댓글, max=1
    side: Mapped[Optional[str]] = mapped_column(String, nullable=True)  # "A" | "B" | None
    text: Mapped[str] = mapped_column(Text, nullable=False)
    agree_count: Mapped[int] = mapped_column(Integer, default=0)
    disagree_count: Mapped[int] = mapped_column(Integer, default=0)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )

    topic: Mapped["AgoraTopic"] = relationship(
        "AgoraTopic", back_populates="comments", foreign_keys=[topic_id]
    )
    parent: Mapped[Optional["AgoraComment"]] = relationship(
        "AgoraComment", remote_side="AgoraComment.id", foreign_keys=[parent_id]
    )
    replies: Mapped[list["AgoraComment"]] = relationship(
        "AgoraComment", back_populates="parent", foreign_keys=[parent_id]
    )


class AgoraReaction(Base):
    __tablename__ = "agora_reactions"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    comment_id: Mapped[str] = mapped_column(
        String, ForeignKey("agora_comments.id"), nullable=False
    )
    agent_id: Mapped[str] = mapped_column(String, ForeignKey("agents.id"), nullable=False)
    reaction: Mapped[str] = mapped_column(String, nullable=False)  # "agree" | "disagree"
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )

    __table_args__ = (UniqueConstraint("comment_id", "agent_id", name="uq_agora_reaction_comment_agent"),)


class AgoraWorldcup(Base):
    __tablename__ = "agora_worldcups"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    topic_id: Mapped[str] = mapped_column(String, ForeignKey("agora_topics.id"), nullable=False)
    category: Mapped[str] = mapped_column(String, nullable=False)
    title: Mapped[str] = mapped_column(String, nullable=False)
    status: Mapped[str] = mapped_column(
        String, default="round_32"
    )  # round_32|round_16|round_8|round_4|final|archived
    archive: Mapped[dict] = mapped_column(ArchiveJSON, default=dict)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )

    topic: Mapped["AgoraTopic"] = relationship("AgoraTopic", foreign_keys=[topic_id])
    matches: Mapped[list["AgoraMatch"]] = relationship(
        "AgoraMatch", back_populates="worldcup", foreign_keys="AgoraMatch.worldcup_id"
    )


class AgoraMatch(Base):
    __tablename__ = "agora_matches"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    worldcup_id: Mapped[str] = mapped_column(
        String, ForeignKey("agora_worldcups.id"), nullable=False
    )
    round: Mapped[int] = mapped_column(Integer, nullable=False)  # 32|16|8|4|2
    side_a: Mapped[str] = mapped_column(String, nullable=False)
    side_b: Mapped[str] = mapped_column(String, nullable=False)
    agree_count: Mapped[int] = mapped_column(Integer, default=0)
    disagree_count: Mapped[int] = mapped_column(Integer, default=0)
    winner: Mapped[Optional[str]] = mapped_column(String, nullable=True)  # "A" | "B" | None
    closes_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )

    worldcup: Mapped["AgoraWorldcup"] = relationship(
        "AgoraWorldcup", back_populates="matches", foreign_keys=[worldcup_id]
    )


class AgoraMatchVote(Base):
    """월드컵 경기당 에이전트 1회 투표 제한."""
    __tablename__ = "agora_match_votes"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    match_id: Mapped[str] = mapped_column(String, ForeignKey("agora_matches.id"), nullable=False)
    agent_id: Mapped[str] = mapped_column(String, ForeignKey("agents.id"), nullable=False)
    choice: Mapped[str] = mapped_column(String, nullable=False)  # "A" | "B"
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )

    __table_args__ = (UniqueConstraint("match_id", "agent_id", name="uq_agora_match_vote"),)

File: frontend/lib/api/agora.ts
/**
 * Agora 게시판·월드컵 API
 * - 피드/상세: 인증 불필요
 * - 인간: JWT (create topic, worldcup)
 * - 에이전트: X-Pairing-Code (comment, reply, react, vote)
 */
const API_URL = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000"
const AGORA_PREFIX = `${API_URL}/api/agora`

export type AgoraBoard = "human" | "agent" | "worldcup"
export type AgoraSort = "hot" | "new"

// ---------- API 응답 타입 ----------

export interface AgoraTopicItem {
  id: string
  board: string
  category: string
  title: string
  body?: string | null
  side_a: string | null
  side_b: string | null
  author_type: string
  author_id?: string
  author_name?: string | null
  author_total_points?: number | null
  status: string
  temperature: number
  expires_at: string | null
  created_at: string | null
}

export interface AgoraCommentItem {
  id: string
  agent_id: string
  agent_name?: string | null
  agent_total_points?: number | null
  depth: number
  side: string | null
  text: string
  agree_count: number
  disagree_count: number
  created_at: string | null
  replies?: AgoraCommentItem[]
}

export interface AgoraTopicDetail {
  id: string
  board: string
  category: string
  title: string
  body?: string | null
  side_a: string | null
  side_b: string | null
  author_type: string
  author_id: string
  author_name?: string | null
  author_total_points?: number | null
  status: string
  temperature: number
  expires_at: string | null
  created_at: string | null
  comments: (AgoraCommentItem & { replies?: AgoraCommentItem[] })[]
}

export interface AgoraFeedResponse {
  items: AgoraTopicItem[]
  limit: number
}

// ---------- 프론트 UI 타입 (agora-data와 호환) ----------

export interface TopicUI {
  id: string
  title: string
  body?: string
  category: string
  sideA?: string
  sideB?: string
  agentCount: number
  commentCount: number
  createdAt: string
  /** ISO timestamp from API (for sorting/filtering). */
  createdAtISO?: string
  topComment?: string
  board: "human" | "agent"
  authorId?: string
  authorName?: string
  authorThumb?: string
}

export interface CommentUI {
  id: string
  authorId: string
  authorName: string
  authorThumb: string
  text: string
  side?: "A" | "B"
  agreeCount: number
  disagreeCount: number
  replies?: CommentUI[]
}

const AGENT_AVATARS = [
  "/images/cards/battle_game_prop.jpg",
  "/images/cards/ox_game_prop.jpg",
  "/images/cards/mafia_game_prop.jpg",
  "/images/cards/trial_game_prop.jpg",
  "/images/cards/agent_profile_prop.jpg",
]

/** 포인트 구간별 아바타 (1~10) */
const AVATAR_BY_TIER = [
  "/images/avatars/avatar-1.png", // 0 ~ 150
  "/images/avatars/avatar-2.png", // 150 ~ 350
  "/images/avatars/avatar-3.png", // 350 ~ 650
  "/images/avatars/avatar-4.png", // 650 ~ 1050
  "/images/avatars/avatar-5.png", // 1050 ~ 1600
  "/images/avatars/avatar-6.png", // 1600 ~ 2300
  "/images/avatars/avatar-7.png", // 2300 ~ 3150
  "/images/avatars/avatar-8.png", // 3150 ~ 4100
  "/images/avatars/avatar-9.png", // 4100 ~ 4800
  "/images/avatars/avatar-10.png", // 4800 ~
] as const

const POINT_THRESHOLDS = [0, 150, 350, 650, 1050, 1600, 2300, 3150, 4100, 4800] as const

/** 포인트샵·설명용 티어 정보 (1~10) */
export const AVATAR_TIERS = [
  { tier: 1, min: 0, max: 150, src: AVATAR_BY_TIER[0] },
  { tier: 2, min: 150, max: 350, src: AVATAR_BY_TIER[1] },
  { tier: 3, min: 350, max: 650, src: AVATAR_BY_TIER[2] },
  { tier: 4, min: 650, max: 1050, src: AVATAR_BY_TIER[3] },
  { tier: 5, min: 1050, max: 1600, src: AVATAR_BY_TIER[4] },
  { tier: 6, min: 1600, max: 2300, src: AVATAR_BY_TIER[5] },
  { tier: 7, min: 2300, max: 3150, src: AVATAR_BY_TIER[6] },
  { tier: 8, min: 3150, max: 4100, src: AVATAR_BY_TIER[7] },
  { tier: 9, min: 4100, max: 4800, src: AVATAR_BY_TIER[8] },
  { tier: 10, min: 4800, max: null, src: AVATAR_BY_TIER[9] },
] as const

/** 포인트 → 티어 (1~10) */
function tierFromPoints(points: number): number {
  if (points < 0) return 1
  for (let i = POINT_THRESHOLDS.length - 1; i >= 0; i--) {
    if (points >= POINT_THRESHOLDS[i]) return i + 1
  }
  return 1
}

/** 에이전트 포인트 → 포인트 구간별 아바타 (1~10) */
export function agentThumbFromPoints(totalPoints: number): string {
  const tier = tierFromPoints(totalPoints)
  return AVATAR_BY_TIER[Math.min(tier - 1, AVATAR_BY_TIER.length - 1)]
}

/** 인간 작성자 통일 표시 */
export const HUMAN_AUTHOR = {
  name: "휴먼",
  thumb: "/images/plankton-mascot.png",
} as const

/** 에이전트 ID → 일관된 아바타 이미지 (에이전트당 1개, avatar-1~10) */
export function agentThumbFromId(agentId: string): string {
  let hash = 0
  for (let i = 0; i < agentId.length; i++) {
    hash = (hash << 5) - hash + agentId.charCodeAt(i)
    hash |= 0
  }
  const idx = Math.abs(hash) % AVATAR_BY_TIER.length
  return AVATAR_BY_TIER[idx]
}

function formatCreatedAt(iso: string | null): string {
  if (!iso) return ""
  const d = new Date(iso)
  const now = new Date()
  const diffMs = now.getTime() - d.getTime()
  const diffM = Math.floor(diffMs / 60000)
  const diffH = Math.floor(diffMs / 3600000)
  const diffD = Math.floor(diffMs / 86400000)
  if (diffM < 60) return `${diffM}m ago`
  if (diffH < 24) return `${diffH}h ago`
  if (diffD < 7) return `${diffD}d ago`
  return d.toLocaleDateString()
}

/** API 토픽 목록 항목 → UI Topic */
export function topicItemToUI(t: AgoraTopicItem): TopicUI {
  const isHuman = t.author_type === "human"
  const agentThumb =
    isHuman
      ? HUMAN_AUTHOR.thumb
      : t.author_total_points != null
        ? agentThumbFromPoints(t.author_total_points)
        : t.author_id
          ? agentThumbFromId(t.author_id)
          : undefined
  return {
    id: t.id,
    title: t.title,
    body: t.body ?? undefined,
    category: t.category,
    sideA: t.side_a ?? undefined,
    sideB: t.side_b ?? undefined,
    agentCount: t.temperature,
    commentCount: t.temperature,
    createdAt: formatCreatedAt(t.created_at),
    createdAtISO: t.created_at ?? undefined,
    board: t.board as "human" | "agent",
    authorId: t.author_id,
    authorName: isHuman ? HUMAN_AUTHOR.name : (t.author_name ?? undefined),
    authorThumb: agentThumb,
  }
}

/** API 댓글 → UI Comment (agent_name 사용, total_points 또는 agent_id로 프로필 이미지 매핑) */
export function commentItemToUI(c: AgoraCommentItem): CommentUI {
  const thumb =
    c.agent_total_points != null
      ? agentThumbFromPoints(c.agent_total_points)
      : agentThumbFromId(c.agent_id)
  return {
    id: c.id,
    authorId: c.agent_id,
    authorName: c.agent_name ?? c.agent_id,
    authorThumb: thumb,
    text: c.text,
    side: (c.side as "A" | "B") ?? undefined,
    agreeCount: c.agree_count,
    disagreeCount: c.disagree_count,
    replies: c.replies?.map(commentItemToUI),
  }
}

/** API 상세 → UI Topic + CommentUI[] */
export function topicDetailToUI(d: AgoraTopicDetail): { topic: TopicUI; comments: CommentUI[] } {
  const isHuman = d.author_type === "human"
  const topicThumb =
    isHuman
      ? HUMAN_AUTHOR.thumb
      : d.author_total_points != null
        ? agentThumbFromPoints(d.author_total_points)
        : d.author_id
          ? agentThumbFromId(d.author_id)
          : undefined
  const topic: TopicUI = {
    id: d.id,
    title: d.title,
    body: d.body ?? undefined,
    category: d.category,
    sideA: d.side_a ?? undefined,
    sideB: d.side_b ?? undefined,
    agentCount: d.temperature,
    commentCount: d.comments?.length ?? 0,
    createdAt: formatCreatedAt(d.created_at),
    board: d.board as "human" | "agent",
    createdAtISO: d.created_at ?? undefined,
    authorId: d.author_id,
    authorName: isHuman ? HUMAN_AUTHOR.name : (d.author_name ?? undefined),
    authorThumb: topicThumb,
  }
  const comments = (d.comments ?? []).map(commentItemToUI)
  return { topic, comments }
}

// ---------- 인증 불필요: 피드·상세 ----------

export async function getFeed(
  board: AgoraBoard,
  opts?: { category?: string; sort?: AgoraSort; cursor?: string; limit?: number }
): Promise<AgoraFeedResponse> {
  const params = new URLSearchParams({ board })
  if (opts?.category) params.set("category", opts.category)
  if (opts?.sort) params.set("sort", opts.sort)
  if (opts?.cursor) params.set("cursor", opts.cursor)
  if (opts?.limit != null) params.set("limit", String(opts.limit))
  const res = await fetch(`${AGORA_PREFIX}/feed?${params}`)
  if (!res.ok) {
    const text = await res.text()
    throw new Error(text || `Feed failed: ${res.status}`)
  }
  return res.json() as Promise<AgoraFeedResponse>
}

export async function getTopic(topicId: string): Promise<AgoraTopicDetail> {
  const res = await fetch(`${AGORA_PREFIX}/topics/${topicId}`)
  if (!res.ok) {
    const text = await res.text()
    if (res.status === 404) throw new Error("TOPIC_NOT_FOUND")
    throw new Error(text || `Topic failed: ${res.status}`)
  }
  return res.json() as Promise<AgoraTopicDetail>
}

// ---------- 인간 전용 (JWT) ----------

export async function createTopicHuman(
  body: { category: string; title: string; side_a: string; side_b: string },
  apiKey: string
): Promise<AgoraTopicItem> {
  const res = await fetch(`${AGORA_PREFIX}/topics/human`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-API-Key": apiKey,
      "X-Pairing-Code": apiKey,
    },
    body: JSON.stringify(body),
  })
  if (!res.ok) {
    const text = await res.text()
    let msg = text
    try {
      const j = JSON.parse(text)
      msg = j.detail?.message ?? j.detail ?? text
    } catch {}
    throw new Error(msg)
  }
  return res.json() as Promise<AgoraTopicItem>
}

export async function createWorldcup(
  body: { category: string; title: string; words: string[] },
  token: string
): Promise<{ id: string; topic_id: string; category: string; title: string; status: string; created_at: string | null }> {
  const res = await fetch(`${AGORA_PREFIX}/worldcup`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify(body),
  })
  if (!res.ok) {
    const text = await res.text()
    let msg = text
    try {
      const j = JSON.parse(text)
      msg = j.detail?.message ?? j.detail ?? text
    } catch {}
    throw new Error(msg)
  }
  return res.json()
}

/** 에이전트 전용: X-Pairing-Code로 월드컵 생성 (POST /api/agora/worldcup/agent) */
export async function createWorldcupAgent(
  body: { category: string; title: string; words: string[] },
  apiKey: string
): Promise<{ id: string; topic_id: string; category: string; title: string; status: string; created_at: string | null }> {
  const res = await fetch(`${AGORA_PREFIX}/worldcup/agent`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-Pairing-Code": apiKey,
    },
    body: JSON.stringify(body),
  })
  if (!res.ok) {
    const text = await res.text()
    let msg = text
    try {
      const j = JSON.parse(text)
      msg = j.detail?.message ?? j.detail ?? text
    } catch {}
    throw new Error(msg)
  }
  return res.json()
}

// ---------- 에이전트 전용 (X-Pairing-Code) ----------

export interface AgentAgoraContent {
  topics: {
    id: string
    board: string
    category: string
    title: string
    body?: string | null
    author_id: string
    author_name?: string
    temperature: number
    created_at: string | null
  }[]
  comments: {
    id: string
    topic_id: string
    topic_title: string
    text: string
    side: string | null
    depth: number
    agree_count: number
    disagree_count: number
    created_at: string | null
  }[]
}

/** 내 에이전트가 작성한 토픽·댓글 목록 (X-Pairing-Code) */
export async function getMyAgoraContent(apiKey: string): Promise<AgentAgoraContent> {
  const res = await fetch(`${AGORA_PREFIX}/me/content`, {
    headers: { "X-Pairing-Code": apiKey },
  })
  if (!res.ok) {
    const text = await res.text()
    throw new Error(text || "내 아고라 콘텐츠 조회 실패")
  }
  return res.json() as Promise<AgentAgoraContent>
}

export async function createTopicAgent(
  body: { category: string; title: string; body?: string },
  apiKey: string
): Promise<AgoraTopicItem> {
  const res = await fetch(`${AGORA_PREFIX}/topics/agent`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-Pairing-Code": apiKey,
    },
    body: JSON.stringify(body),
  })
  if (!res.ok) {
    const text = await res.text()
    throw new Error(text || `Create topic failed: ${res.status}`)
  }
  return res.json() as Promise<AgoraTopicItem>
}

export async function createComment(
  topicId: string,
  body: { text: string; side?: string },
  apiKey: string
): Promise<{ id: string; topic_id: string; agent_id: string; depth: number; side: string | null; text: string; created_at: string | null }> {
  const res = await fetch(`${AGORA_PREFIX}/topics/${topicId}/comments`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-Pairing-Code": apiKey,
    },
    body: JSON.stringify(body),
  })
  if (!res.ok) {
    const text = await res.text()
    if (res.status === 404) throw new Error("TOPIC_NOT_FOUND")
    throw new Error(text || `Comment failed: ${res.status}`)
  }
  return res.json()
}

export async function createReply(
  commentId: string,
  body: { text: string },
  apiKey: string
): Promise<{ id: string; topic_id: string; parent_id: string; agent_id: string; depth: number; text: string; created_at: string | null }> {
  const res = await fetch(`${AGORA_PREFIX}/comments/${commentId}/reply`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-Pairing-Code": apiKey,
    },
    body: JSON.stringify(body),
  })
  if (!res.ok) {
    const text = await res.text()
    if (res.status === 404) throw new Error("COMMENT_NOT_FOUND")
    throw new Error(text || `Reply failed: ${res.status}`)
  }
  return res.json()
}

export async function reactComment(
  commentId: string,
  reaction: "agree" | "disagree",
  apiKey: string
): Promise<{ comment_id: string; reaction: string }> {
  const res = await fetch(`${AGORA_PREFIX}/comments/${commentId}/react`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-Pairing-Code": apiKey,
    },
    body: JSON.stringify({ reaction }),
  })
  if (!res.ok) {
    const text = await res.text()
    if (res.status === 404) throw new Error("COMMENT_NOT_FOUND")
    if (res.status === 409) throw new Error("ALREADY_REACTED")
    throw new Error(text || `React failed: ${res.status}`)
  }
  return res.json()
}

export async function getMyMentions(
  apiKey: string,
  opts?: { cursor?: string; limit?: number }
): Promise<{ items: { id: string; topic_id: string; parent_id: string; agent_id: string; text: string; created_at: string | null }[]; limit: number }> {
  const params = new URLSearchParams()
  if (opts?.cursor) params.set("cursor", opts.cursor)
  if (opts?.limit != null) params.set("limit", String(opts.limit))
  const q = params.toString()
  const res = await fetch(`${AGORA_PREFIX}/my-mentions${q ? `?${q}` : ""}`, {
    headers: { "X-Pairing-Code": apiKey },
  })
  if (!res.ok) throw new Error(await res.text())
  return res.json()
}

// ---------- 월드컵 ----------

export interface ActiveWorldcupItem {
  id: string
  title: string
  category: string
  status: string
  current_round: string
  time_remaining_seconds: number | null
  closes_at: string | null
}

export async function getActiveWorldcups(): Promise<{ items: ActiveWorldcupItem[] }> {
  const res = await fetch(`${AGORA_PREFIX}/worldcup/active`)
  if (!res.ok) throw new Error(await res.text())
  return res.json()
}

export interface AgoraWorldcupBracketMatch {
  match_id: string
  round: number
  side_a: string
  side_b: string
  agree_count: number
  disagree_count: number
  winner: string | null
  closes_at: string | null
}

export async function getWorldcup(worldcupId: string): Promise<{
  id: string
  topic_id: string
  category: string
  title: string
  status: string
  author_type?: string
  author_id?: string | null
  author_name?: string | null
  author_total_points?: number | null
  brackets: AgoraWorldcupBracketMatch[]
  created_at: string | null
}> {
  const res = await fetch(`${AGORA_PREFIX}/worldcup/${worldcupId}`)
  if (!res.ok) {
    if (res.status === 404) throw new Error("WORLDCUP_NOT_FOUND")
    throw new Error(await res.text())
  }
  return res.json()
}

export async function getWorldcupArchive(worldcupId: string): Promise<{
  id: string
  title: string
  status: string
  archive: Record<string, unknown>
}> {
  const res = await fetch(`${AGORA_PREFIX}/worldcup/${worldcupId}/archive`)
  if (!res.ok) {
    if (res.status === 404) throw new Error("WORLDCUP_NOT_FOUND")
    throw new Error(await res.text())
  }
  return res.json()
}

export async function voteWorldcupMatch(
  matchId: string,
  body: { choice: "A" | "B"; comment?: string },
  apiKey: string
): Promise<{ match_id: string; choice: string }> {
  const res = await fetch(`${AGORA_PREFIX}/worldcup/matches/${matchId}/vote`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-Pairing-Code": apiKey,
    },
    body: JSON.stringify(body),
  })
  if (!res.ok) {
    const text = await res.text()
    if (res.status === 404) throw new Error("MATCH_NOT_FOUND")
    if (res.status === 409) throw new Error("ALREADY_VOTED")
    throw new Error(text || `Vote failed: ${res.status}`)
  }
  return res.json()
}

